File: Homework/Mary_Antonyan/C++/30_03_2017/Chess/bishop.cpp
#include "bishop.hpp"
#include "board.hpp"
#include <cmath>
#include <iostream>

//Constructor
Chess::Bishop::Bishop(const Chess::Figure::Chessman figure, 
                        const Chess::Figure::Color color, 
                        const int row, const int coloumn) 
    : Figure(figure, color, row, coloumn) 
{}

//Destuctor
Chess::Bishop::~Bishop() {}

//Method override
void Chess::Bishop::step(const int row, const int coloumn) {
    if (0 <= row && 8 > row && 0 <= coloumn && 8 > row) { 
        if (abs(row - coloumn) == abs(mRow - mColoumn) && 
            (Chess::board[row][coloumn] == nullptr || 
             Chess::board[row][coloumn]->getColor() != mColor)) {
            std::cout << "Move to [" << row << " , " << coloumn << "]\n";
            int tmpRow = mRow;
            int tmpCol = mColoumn;
            Chess::board[row][coloumn] = new Bishop(mFigure, mColor, row, coloumn);
            Chess::board[tmpRow][tmpCol] = nullptr;
        } else {
            std::cout << "Can't move to [" << row << " , " << coloumn << "]\n";
        }
    } else {
        std::cout << "Out of board" << std::endl;
    }
}

File: Homework/Khachik/Java/08-08/chess/figures/Bishop.java
package figures;

public class Bishop extends Figure {
    public Bishop(int x, int y, char color, char name){
        super(x, y, color, name);
    }
    public boolean step(int x, int y, Figure[][] fig) {
        if(0 <= x && 7 >= x && 0 <= y && 7 >= y) {
        if (Math.abs(this.getx() - x) == Math.abs(this.gety() - y)) {
            if (this.getx() < x && this.gety() < y) {
                for (int i = this.getx() + 1,  j = this.gety() + 1; i < x && j < y; ++i, ++j) {
                    if(null != fig[i][j]) {
                        return false;
                    }
                }
            }
            if (this.getx() < x && this.gety() > y) {
                for (int i = this.getx() + 1, j = this.gety() - 1; i < x && j > y; ++i, --j) {
                    if(null != fig[i][j]) {
                        return false;
                    }
                }
            }
            if (this.getx() > x && this.gety() > y) {
                for (int i = this.getx() - 1, j = this.gety() - 1; i > x && j > y; --i, --j) {
                    if(null != fig[i][j]) {
                        return false;
                    }
                }
            }
            if (this.getx() > x && this.gety() > y) {
                for (int i = this.getx() - 1, j = this.gety() + 1; i > x && j < y; --i, ++j) {
                    if(null != fig[i][j]) {
                        return false;
                    }
                }
            }
            if (null == fig[x][y]) {
                return true;
            } else if (fig[x][y].getcolor() != this.getcolor()){
                fig[x][y] = null;
                return true;
            }
        }       
    }
    return false;   
    }
} 